src/library/scala/collection/IterableOnce.scala

```diff
@@ -700,153 +701,171 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
+   */
   def reduce[B >: A](op: (B, B) => B): B = reduceLeft(op)
 
   /** Reduces the elements of this $coll, if any, using the specified
-    *  associative binary operator.
-    *
-    *  $undefinedorder
-    *
-    *  @tparam B     A type parameter for the binary operator, a supertype of `A`.
-    *  @param op      A binary operator that must be associative.
-    *  @return        An option value containing result of applying reduce operator `op` between all
-    *                 the elements if the collection is nonempty, and `None` otherwise.
-    */
+   *  associative binary operator.
+   *
+   *  $undefinedorder
+   *
+   *  @tparam B     A type parameter for the binary operator, a supertype of `A`.
+   *  @param op      A binary operator that must be associative.
+   *  @return        An option value containing result of applying reduce operator `op` between all
+   *                 the elements if the collection is nonempty, and `None` otherwise.
+   */
   def reduceOption[B >: A](op: (B, B) => B): Option[B] = reduceLeftOption(op)
 
   /** Applies a binary operator to all elements of this $coll,
-    *  going left to right.
-    *  $willNotTerminateInf
-    *  $orderDependentFold
-    *
-    *  @param  op    the binary operator.
-    *  @tparam  B    the result type of the binary operator.
-    *  @return  the result of inserting `op` between consecutive elements of this $coll,
-    *           going left to right:
-    *           `op( op( ... op(x,,1,,, x,,2,,) ..., x,,n-1,,), x,,n,,)` where `x,,1,,, ..., x,,n,,`
-    *           are the elements of this $coll.
-    *  @throws UnsupportedOperationException if this $coll is empty.   */
+   *  going left to right.
+   *  $willNotTerminateInf
+   *  $orderDependentFold
+   *
+   *  @param  op    the binary operator.
+   *  @tparam  B    the result type of the binary operator.
+   *  @return  the result of inserting `op` between consecutive elements of this $coll,
+   *           going left to right:
+   *           `op( op( ... op(x,,1,,, x,,2,,) ..., x,,n-1,,), x,,n,,)` where `x,,1,,, ..., x,,n,,`
+   *           are the elements of this $coll.
+   *  @throws UnsupportedOperationException if this $coll is empty.
+   */
   def reduceLeft[B >: A](op: (B, A) => B): B = this match {
     case seq: IndexedSeq[A @unchecked] if seq.length > 0 => foldl(seq, 1, seq(0), op)
-    case _ =>
-      val it = iterator
-      if (it.isEmpty)
-        throw new UnsupportedOperationException("empty.reduceLeft")
-
-      var first = true
-      var acc: B = null.asInstanceOf[B]
-
-      while (it.hasNext) {
-        val x = it.next()
-        if (first) {
-          acc = x
-          first = false
-        }
-        else acc = op(acc, x)
-      }
+    case _ if knownSize == 0 => throw new UnsupportedOperationException("empty.reduceLeft")
+    case _ => reduceLeftIterator[B](throw new UnsupportedOperationException("empty.reduceLeft"))(op)
```

Comment:
Might be worth handling differently case where `knownSize == -1` and `knownSize != -1`. If the `knownSize` is "known", then we can leverage that knowledge to avoid calls to `hasNext` which might be expensive. 

```scala
val it = iterator
val ks = knownSize
if (ks == -1) {
  while (it.hasNext) {
    // ...
  }
else {
  var i = 0
  while(i < ks) {
    //...
    i += 1
  }
}
```

Reply:
(It very well may not be worth it though)

Reply:
That's an interesting idea. hasNext doesn't have to be much more expensive.

Reply:
Actually, I recently coined "knownSize inconsistent with iterator" https://github.com/scala/scala/pull/10194

Iterators can fail to deliver, perhaps.

Reply:
May want to override also for LinearSeq to use `head` and `tail` instead of `iterator`

Reply:
This kind of talk must have inspired me to tweak `sameElements` for case of both known sizes are zero.